onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -361,32 +363,37 @@ void UpsampleBilinear(int64_t batch_size,
     }
   }
 
-  for (int64_t n = 0; n < batch_size; ++n) {
-    for (int64_t c = 0; c < num_channels; ++c) {
-      for (int64_t y = 0; y < output_height; ++y) {
-        for (int64_t x = 0; x < output_width; ++x) {
-          // when use_extrapolation is set and original index of x or y is out of the dim range
-          // then use extrapolation_value as the output value.
-          if (use_extrapolation &&
-              ((y_original[y] < 0 || y_original[y] > static_cast<float>(input_height - 1)) ||
-               (x_original[x] < 0 || x_original[x] > static_cast<float>(input_width - 1)))) {
-            Ydata[output_width * y + x] = static_cast<T>(extrapolation_value);
-            continue;
-          }
+  const auto interpolate = [&](int32_t index) {
+    const T* X = Xdata + index * input_height * input_width;
+    T* Y = Ydata + index * output_height * output_width;
+    for (int64_t y = 0; y < output_height; ++y) {
+      for (int64_t x = 0; x < output_width; ++x) {
+        // when use_extrapolation is set and original index of x or y is out of the dim range
+        // then use extrapolation_value as the output value.
+        if (use_extrapolation &&
+            ((y_original[y] < 0 || y_original[y] > static_cast<float>(input_height - 1)) ||
+            (x_original[x] < 0 || x_original[x] > static_cast<float>(input_width - 1)))) {
+          Y[output_width * y + x] = static_cast<T>(extrapolation_value);
+          continue;
+        }
 
-          T X11 = Xdata[input_width_mul_y1[y] + in_x1[x]];
-          T X21 = Xdata[input_width_mul_y1[y] + in_x2[x]];
-          T X12 = Xdata[input_width_mul_y2[y] + in_x1[x]];
-          T X22 = Xdata[input_width_mul_y2[y] + in_x2[x]];
+        T X11 = X[input_width_mul_y1[y] + in_x1[x]];
+        T X21 = X[input_width_mul_y1[y] + in_x2[x]];
+        T X12 = X[input_width_mul_y2[y] + in_x1[x]];
+        T X22 = X[input_width_mul_y2[y] + in_x2[x]];
 
-          Ydata[output_width * y + x] = static_cast<T>(dx2[x] * dy2[y] * X11 +
-                                                       dx1[x] * dy2[y] * X21 +
-                                                       dx2[x] * dy1[y] * X12 +
-                                                       dx1[x] * dy1[y] * X22);
-        }
+        Y[output_width * y + x] = static_cast<T>(dx2[x] * dy2[y] * X11 +
+                                                    dx1[x] * dy2[y] * X21 +
+                                                    dx2[x] * dy1[y] * X12 +
+                                                    dx1[x] * dy1[y] * X22);
       }
-      Xdata += input_height * input_width;
-      Ydata += output_width * output_height;
+    }
+  };
+  if (thread_pool) {
+    thread_pool->ParallelFor(static_cast<int32_t>(batch_size * num_channels), interpolate);
+  } else {
```

Comment:
Did you try using BatchParallelFor to see if perf would be further improved by doing batches in each thread instead of a single index at a time?

Reply:
Didn't try it. I felt like having the entire channel as a unit of work was enough that the overhead of scheduling each channel wouldn't be too bad, but of course it depends on the image dimensions.

I talked to Changming about it before opening the PR. He said this looked like a good improvement for now and more advanced benchmarking and optimization could be discussed later.

Reply:
The new code can perform bad only when output_height*output_width is small.  There are rooms to improve it but I think so far it is ok. At least it unblocked the use case so that we can move forward and continuously improve it.